### DISCUSSION TITLE
Assets served from incorrect directory when 404 error occurs

### DISCUSSION THREAD

**Discussion by Kodlak15** [tokio-rs/axum#1946] [Q&A] [answered]
I am attempting to serve a SvelteKit app using Axum, and am having an issue with my handling of 404 errors. The build files for my Svelte app are located in the `client/build` directory, and the assets needed to render the pages properly (CSS, JS, etc.) are located in the `client/build/_app` directory. I would like for `client/build/404.html` to be served when a page not found error occurs. 

```
let serve_dir = get_service( ServeDir::new("client/build").not_found_service(ServeFile::new("client/build/404.html")),
)
.handle_error(handle_error);

let static_files = Router::new()
    .nest_service("/", serve_dir.clone())
    .route("/my_route", get(my_route));
```

To be clear, my code is doing exactly that. The issue is that when a 404 error occurs, the assets are pulled from the parent directory of the path that caused the 404 error to occur rather than `client/build/_app`.

```
# Case 1: localhost/aaa -> 404 error occurs, assets fetched from from /_app/...
# Case 2: localhost/aaa/bbb -> 404 error occurs, assets fetched from /aaa/_app/...
```

In both of these cases `404.html` is correctly served. In case 1 everything works as I would expect. However, in case 2 all of the styling is broken because assets are being fetched from `client/build/aaa/_app`, which does not exist, instead of `client/build/_app`. It is also worth mentioning that if `localhost/aaa/bbb` is a legitimate path this is a non issue. The issue is only occurring when a 404 error happens and the path is 2 or more levels below the root of the project. I am convinced that this is an issue with my Rust code, as the issue is not occuring when I run `npm run preview` to preview my Svelte build. I am simply hoping that somebody here who is more knowledgeable than I can shed some light on why this issue may be occurring, and what I can do to fix it. Thanks in advance for taking the time to read this.

**Comment by davidpdrsn** [comment, marked as answer]
How are asset paths specified in 404.html? If you have an asset at `GET /style.css` it doesn't matter where the file that's requesting that is because it's an absolute path.

**Reply by Kodlak15** [reply]
Apologies for the slightly delayed response. To answer your question, the asset paths in 404.html are relative paths. I went ahead and manually made the paths absolute paths instead, and everything works as expected now. I am going to mark this as the answer as this seems to indicate that the issue pertains to how my Svelte files are being built. I appreciate your help!
